src/core/instance/inject.js

```diff
@@ -57,6 +57,8 @@ export function resolveInject (inject: any, vm: Component): ?Object {
         }
         source = source.$parent
       }
+      process.env.NODE_ENV !== 'production' && typeof result[key] === 'undefined' &&
```

Comment:
Thanks! I added a test case and a solution using`Object.keys`.

In the end solution is not based on `result.hasOwnProperty(key)` or a `result.hasOwnProperty && !result.hasOwnProperty(key)` because they failed some tests, not sure why. Also Flow was reporting strange error, which required a change in var definition to an odd: `const result = (Object.create(null): Object)`.